substreams/src/parser.rs

```diff
@@ -0,0 +1,53 @@
+use pest::{iterators::Pair, Parser};
+use pest_derive::Parser;
+
+#[derive(Parser)]
+#[grammar = "parser_rule.pest"]
+pub struct EParser;
+
+pub fn evaluate_expression(keys: Vec<String>, input: &str) -> bool {
+    let successful_parse = EParser::parse(Rule::expression, input).unwrap();
+    return evaluate_rule(successful_parse.clone().into_iter().next().unwrap(), keys);
```

Comment:
Let's avoid all `.clone()` in your code, there is a good chance that we can remove all of them.